File: frontend/de/unika/ipd/grgen/ir/model/type/EnumType.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 6.5
 * Copyright (C) 2003-2022 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

/**
 * @author Sebastian Hack
 */

package de.unika.ipd.grgen.ir.model.type;

import java.util.Collections;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

import de.unika.ipd.grgen.ir.*;
import de.unika.ipd.grgen.ir.model.EnumItem;
import de.unika.ipd.grgen.ir.type.basic.PrimitiveType;

/**
 * An enumeration type.
 */
public class EnumType extends PrimitiveType implements ContainedInPackage
{
	private String packageContainedIn;

	private final List<EnumItem> items = new LinkedList<EnumItem>();

	/** Make a new enum type.
	 *  @param ident The identifier of this enumeration. */
	public EnumType(Ident ident)
	{
		super("enum type", ident);
	}

	/** Add teh given item to a this enum type and autoenumerate it. */
	public void addItem(EnumItem item)
	{
		items.add(item);
	}

	/** @return A list with the identifiers in the enum type. */
	public List<EnumItem> getItems()
	{
		return Collections.unmodifiableList(items);
	}

	/** @see de.unika.ipd.grgen.ir.type.Type#classify() */
	@Override
	public TypeClass classify()
	{
		return TypeClass.IS_INTEGER;
	}

	@Override
	public String getPackageContainedIn()
	{
		return packageContainedIn;
	}

	public void setPackageContainedIn(String packageContainedIn)
	{
		this.packageContainedIn = packageContainedIn;
	}

	@Override
	protected void canonicalizeLocal()
	{
		Collections.sort(items, Identifiable.COMPARATOR);
	}

	@Override
	public void addToDigest(StringBuffer sb)
	{
		sb.append(this);
		sb.append('[');

		int i = 0;
		for(Iterator<EnumItem> it = items.iterator(); it.hasNext(); i++) {
			EnumItem ent = it.next();
			if(i > 0)
				sb.append(',');
			sb.append(ent);
		}

		sb.append(']');
	}
}

File: frontend/de/unika/ipd/grgen/ir/model/type/EnumType.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

/// <summary>
/// @author Sebastian Hack
/// </summary>

namespace de.unika.ipd.grgen.ir.model.type
{

using System.Collections.Generic;
using System.Text;

using de.unika.ipd.grgen.ir;
using EnumItem = de.unika.ipd.grgen.ir.model.EnumItem;
using PrimitiveType = de.unika.ipd.grgen.ir.type.basic.PrimitiveType;

/// <summary>
/// An enumeration type.
/// </summary>
public class EnumType : PrimitiveType, ContainedInPackage
{
	private string packageContainedIn;

	private readonly List<EnumItem> items = new List<EnumItem>();

	/// <summary>
	/// Make a new enum type. </summary>
	///  <param name="ident"> The identifier of this enumeration.  </param>
	public EnumType(Ident ident)
		: base("enum type", ident)
	{
	}

	/// <summary>
	/// Add teh given item to a this enum type and autoenumerate it. </summary>
	public virtual void AddItem(EnumItem item)
	{
		items.Add(item);
	}

	/// <returns> A list with the identifiers in the enum type. </returns>
	public virtual IList<EnumItem> Items
	{
		get
		{
		return items.AsReadOnly();
		}
	}

	/// <seealso cref="de.unika.ipd.grgen.ir.type.Type.classify() "/>
	public override TypeClass Classify()
	{
		return TypeClass.IS_INTEGER;
	}

	public virtual string PackageContainedIn
	{
		get
		{
		return packageContainedIn;
		}
		set
		{
		this.packageContainedIn = value;
		}
	}


	protected internal override void CanonicalizeLocal()
	{
		items.Sort(Identifiable.COMPARATOR);
	}

	public override void AddToDigest(StringBuilder sb)
	{
		sb.Append(this);
		sb.Append('[');

		int i = 0;
		foreach(EnumItem ent in items)
		{
			if(i > 0)
				sb.Append(',');
			sb.Append(ent);
			++i;
		}

		sb.Append(']');
	}
}

}
